File: ruby/test/wire/environments_test.rb
# frozen_string_literal: true

require_relative "wiremock_test_case"

class EnvironmentsWireTest < WireMockTestCase
  def setup
    super

    @client = Sikaru::Client.new(
      api_key: "<token>",
      base_url: WIREMOCK_BASE_URL
    )
  end

  def test_environments_list_managed_environments_with_wiremock
    test_id = "environments.list_managed_environments.0"

    @client.environments.list_managed_environments(
      project_id: "project_id",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "environments.list_managed_environments.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/projects/project_id/environments",
      query_params: nil,
      expected: 1
    )
  end

  def test_environments_create_managed_environment_with_wiremock
    test_id = "environments.create_managed_environment.0"

    @client.environments.create_managed_environment(
      project_id: "project_id",
      environment_slug: "environmentSlug",
      provider_type: "providerType",
      runtime_provider: "runtimeProvider",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "environments.create_managed_environment.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/projects/project_id/environments",
      query_params: nil,
      expected: 1
    )
  end
end

File: rust/tests/environments_test.rs
use sikaru::prelude::*;

mod wire_test_utils;

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_environments_list_managed_environments_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .environments
        .list_managed_environments(&"project_id".to_string(), None)
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("GET", "/v1/projects/project_id/environments", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_environments_create_managed_environment_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .environments
        .create_managed_environment(
            &"project_id".to_string(),
            &CreateManagedEnvironmentRequest {
                environment_slug: "environmentSlug".to_string(),
                provider_type: "providerType".to_string(),
                runtime_provider: "runtimeProvider".to_string(),
                config_refs: None,
                status: None,
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/v1/projects/project_id/environments", None, 1)
        .await
        .unwrap();
}
